### PR TITLE
add Hyper-parameter Optimization algorithm

### PR DESCRIPTION
**Pull request by n1ck-guo**
## Type of Change

add new feature

## Description

Add new hpo algorithm, include search algorithms, simulated annealing and search space.

## Expected Behavior & Potential Risk

no

## How has this PR been tested?

how to reproduce the test (including hardware information)

## Dependency Change?

xgboost

### CODE REVIEW

File: examples/pytorch/nlp/huggingface_models/text-classification/pruning/hpo/run_glue_no_trainer.py [hunk 1]
```diff
@@ -0,0 +1,669 @@
...
+        eval_metric = metric.compute()
+        logger.info(f"mnli-mm: {eval_metric}")
+    return eval_metric
+
+
+if __name__ == "__main__":
+#     main()
+    os.environ['CUDA_VISIBLE_DEVICES'] = '1'
+    import time
+    from neural_compressor.compression.hpo import XgbSearcher, DiscreteSearchSpace, ContinuousSearchSpace
```
Comment (wenhuach21) [inline examples/pytorch/nlp/huggingface_models/text-classification/pruning/hpo/run_glue_no_trainer.py:640]: is it better to  combine DiscreteSearchSpace and ContinuousSearchSpace ?

Comment (wenhuach21) [reply examples/pytorch/nlp/huggingface_models/text-classification/pruning/hpo/run_glue_no_trainer.py:640]: the style of inc is to wrap all the stuff in one API, please hold on this pr, we need to carefully design the interface

Comment (n1ck-guo) [reply examples/pytorch/nlp/huggingface_models/text-classification/pruning/hpo/run_glue_no_trainer.py:640]: done

File: neural_compressor/compression/hpo/search_space.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+        bound=None,
+        interval=None,
+        value=None,
+        type=None
+        ):
+    """Get registered searchspace class.
+
+    Args:
+        bound: A tuple or list that limit the max and min number of search space.
+        interval: Only for descrete search space. Intervals in discrete space.
```
Comment (wenhuach21) [inline neural_compressor/compression/hpo/search_space.py:32]: give an example here

File: docs/source/pruning.md [hunk 3]
```diff
@@ -386,6 +386,37 @@ Please refer to [pruning examples](../../examples/README.md#Pruning-1) for more
...
+hyper-parameters, especially for large datasets or complex ML algorithms with a large number of
+hyperparameters.
+2. It improves the performance of ML models. Many ML hyper-parameters have different optimums to
+achieve best performance in different datasets or problems.
+3. It makes the models and research more reproducible. Only when the same level of hyper-parameter
+tuning process is implemented can different ML algorithms be compared fairly; hence, using a same HPO
+method on different ML algorithms also helps to determine the most suitable ML model for a specific
+problem
+
+### Get Started with HPO API
```
Comment (wenhuach21) [inline docs/source/pruning.md:401]: move this readme to the HPO folder and add a concise summary in pruning.md. there are two pruning.md, the other one is in the pruner folder, please help sync them

File: neural_compressor/compression/hpo/search_space.py (hunk 2 above)
Comment (n1ck-guo) [reply neural_compressor/compression/hpo/search_space.py:28]: done

File: docs/source/pruning.md (hunk 3 above)
Comment (n1ck-guo) [reply docs/source/pruning.md]: done

File: docs/source/pruning.md [hunk 4]
```diff
@@ -52,8 +52,9 @@ Pruning
 
 4. [Sparse Model Deployment](#sparse-model-deployment)
 
+5. [Pruning With HPO](#pruning-with-hyperparameter-optimization)
 
-5. [Reference](#reference)
+6. [Reference](#reference)
```
Comment (wenhuach21) [inline docs/source/pruning.md:59]: penghui updated this one before but not pruner/readme, please help sync.

**Review by wenhuach21 — APPROVED** [review]

**Review by WeiweiZhang1 — APPROVED** [review]

**Pull request merged**